firebase-firestore/src/main/java/com/google/firebase/firestore/util/ProtoValueUtil.java

```diff
@@ -0,0 +1,266 @@
+    Iterator<Map.Entry<String, Value>> iterator2 =
+        new TreeMap<>(right.getMapValue().getFieldsMap()).entrySet().iterator();
+    while (iterator1.hasNext() && iterator2.hasNext()) {
+      Map.Entry<String, Value> entry1 = iterator1.next();
+      Map.Entry<String, Value> entry2 = iterator2.next();
+      int keyCompare = entry1.getKey().compareTo(entry2.getKey());
+      if (keyCompare != 0) {
+        return keyCompare;
+      }
+      int valueCompare = compare(entry1.getValue(), entry2.getValue());
+      if (valueCompare != 0) {
+        return valueCompare;
+      }
+    }
+
+    // Only equal if both iterators are exhausted.
+    return Util.compareBooleans(iterator1.hasNext(), iterator2.hasNext());
+  }
+
+  private static int compareArrays(Value left, Value right) {
+    int minLength =
+        Math.min(left.getArrayValue().getValuesCount(), right.getArrayValue().getValuesCount());
+    for (int i = 0; i < minLength; i++) {
+      int cmp = compare(left.getArrayValue().getValues(i), right.getArrayValue().getValues(i));
+      if (cmp != 0) {
+        return cmp;
+      }
+    }
+    return Util.compareIntegers(
+        left.getArrayValue().getValuesCount(), right.getArrayValue().getValuesCount());
+  }
+
+  private static int compareGeoPoints(Value left, Value right) {
+    int comparison =
+        Util.compareDoubles(
+            left.getGeoPointValue().getLatitude(), right.getGeoPointValue().getLatitude());
+    if (comparison == 0) {
+      return Util.compareDoubles(
+          left.getGeoPointValue().getLongitude(), right.getGeoPointValue().getLongitude());
+    }
+    return comparison;
+  }
+
+  private static int compareReferences(Value left, Value right) {
+    List<String> leftSegments = Splitter.on('/').splitToList(left.getReferenceValue());
+    List<String> rightSegments = Splitter.on('/').splitToList(right.getReferenceValue());
+    int minLength = Math.min(leftSegments.size(), rightSegments.size());
+    for (int i = 0; i < minLength; i++) {
+      int cmp = leftSegments.get(i).compareTo(rightSegments.get(i));
+      if (cmp != 0) {
+        return cmp;
+      }
+    }
+    return Util.compareIntegers(leftSegments.size(), rightSegments.size());
+  }
+
+  private static int compareTimestamps(Value left, Value right) {
+    if (left.getTimestampValue().getSeconds() == right.getTimestampValue().getSeconds()) {
+      return Integer.signum(
+          left.getTimestampValue().getNanos() - right.getTimestampValue().getNanos());
+    }
+    return Long.signum(
+        left.getTimestampValue().getSeconds() - right.getTimestampValue().getSeconds());
+  }
+
+  private static int compareNumbers(Value left, Value right) {
+    if (left.getValueTypeCase() == Value.ValueTypeCase.DOUBLE_VALUE) {
+      double thisDouble = left.getDoubleValue();
+      if (right.getValueTypeCase() == Value.ValueTypeCase.DOUBLE_VALUE) {
+        return Util.compareDoubles(thisDouble, right.getDoubleValue());
+      } else {
```

Comment:
You could express this with less code if you handled all the unexpected cases in one place, below:

```
if (left is double) {
  if (right is double) {
    return ...;
  } else if (right is int) {
    return ...;
  }
} else if (left is int) {
  if (right is double) {
    return ...;
  } else if (right is int) {
    return ...;
  }
}

hardAssert("Unexpected types %s and %s, left, right);
```

Reply:
Great suggestion. Updated.